src/object/cli_obj.c

```diff
@@ -1984,7 +2109,10 @@ obj_punch_internal(tse_task_t *api_task, enum obj_rpc_opc opc,
 			goto out_task;
 
 		args = tse_task_buf_embedded(task, sizeof(*args));
-		shard			= shard_first + i;
+		if (shard_first != &shard_first_inline)
+			shard		= shard_first[i];
+		else
+			shard		= shard_first_inline + i;
```

Comment:
although I'm not very sure what's the real problem met by Mohamod's test, it seems to me it is related with that for OBJ_PUNCH needs to punch all shards, current code will select one as leader.
For some special obj layout, for example the LARGE_RW object class, it will use all targets. When there are multiple VOS tgts in one node, for example just with one server and 8 vos on it. the shard_nr is 8, and select one as leader, but the problem is that all other 7 non-leader shards actually are all on the same node. Seems in that case DTX will meet some problems? (punch req forward to leader itself multiple times).
Not only LARGE_RW object class, some other layout/placement possibly cause multiple RDGs on same node, if that node select as leader also with that problem.
So a fix maybe that check if some non-leader shards is on the same node as leader, for that may need some special handling for OBJ_PUNCH.
seems this patch is not really helpful to resolve it, but maybe I totally misunderstood it?

Reply:
Your analysis is correct. We hit trouble when punch the object that has multiple stripes which belong to different redundancy groups. Currently, DTX only can handle the consistency within single redundancy group. But the stripes in this case belong to different redundancy groups. So sending OBJ_PUNCH RPC to single 'leader' will confuse the DTX logic as to the 'leader' cannot control other stripes to commit related DTX.

This patch makes the client to send OBJ_PUNCH RPC to the leaders for all related redundancy groups. For each redundancy group, its leader only handle its own DTX. That is the logic for this patch. As explained in the patch comment, current patch is temporary solution, there is no guarantee that all related redundancy groups can punch related stripe successfully. So under some bad cases, the object may be partially punched. The issue will be resolved via DAOS distributed transaction in the future.

Reply:
thanks for explanation, but seems still some problems.
I think the reason why this patch resolved mohamod's infinite retry issue is in his case for example one server with 8 VOS (LARGE RW object class), there is 8 targets in the object's placement map - 8 RDGs (only one tgt for each RDG). After this patch will send 8 OBJ_PUNCH RPCs to the same server, each RPC carried fw_tgts is NULL so it will not be forwarded/dispatched again at servers-side, so aovided the infinite DER_INPROGESS+retry.

But how about some other bad cases for example:
still one server with 8 VOS, but somehow in object class set grp_nr = 4 and replicas = 2, then there are 4 RDGs and within each RDG with 2 tgts. After this patch client will send 4 OBJ_PUNCH RPC to server but for each RPC server still will forward the RPC to another non-leader shrad/tgt in that RDG. non-leader and leader on the same server/node again, seems still with cause infinite DER_INPROGESS+retry?

it seems to me that we still need some logic to check that if non-leader and leader on same server? or seems a simpler fix could that at server-side check if the non-leader shards is on the same server of leader shard in that case add some special handling? then probably need not change client-side logical for now, and leave the multiple RDG things to future?

Reply:
I think we should NOT allow multiple replicas to reside on the same server, right? Because it is meaningless if support that.

Reply:
Ah yes, just checked there is "if (rop->rop_grp_size > rimap->rmp_domain_nr) return EINVAL".
Then your patch seems find to resolve it for now. sorry for noise.

Reply:
Could you please the case when we could encounter the issue where:
"So under some bad cases, the object may be partially punched."
Thanks.

Reply:
It will happen if some redundancy groups destroy their stripes successfully, but some other redundancy groups failed to do that, then part of stripes gone away but others still there. As for why some redundancy groups failed to that, it may cause kinds of reason, such as node failure, or client failure when only sent some OBJ PUNCH RPCs.

Reply:
I see, thanks for the clarification